File: python/src/runapi/suno/resources/regenerate_validation_phrase.py
"""Suno regenerate_validation_phrase resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource

from .. import _validators
from ..types import CompletedValidationPhraseResponse, ValidationPhraseResponse


class RegenerateValidationPhrase(Resource):
    """Regenerate a validation phrase for a prior task."""

    ENDPOINT = "/api/v1/suno/regenerate_validation_phrase"

    RESPONSE_CLASS = ValidationPhraseResponse
    COMPLETED_RESPONSE_CLASS = CompletedValidationPhraseResponse

    def run(self, **params: Any) -> Any:
        """Request a new validation phrase and poll until it completes.

        Args:
            **params: validation-phrase parameters.

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create a validation-phrase task and return immediately with an id.

        Args:
            **params: validation-phrase parameters.

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of a validation-phrase task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        _validators.validate_regenerate_validation_phrase(params)

File: ruby/lib/runapi/suno/resources/regenerate_validation_phrase.rb
# frozen_string_literal: true

module RunApi
  module Suno
    module Resources
      # Step 2 (optional) of voice cloning: requests a new, easier validation phrase.
      class RegenerateValidationPhrase
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/suno/regenerate_validation_phrase"
        RESPONSE_CLASS = Types::ValidationPhraseResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedValidationPhraseResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          Validators.validate_regenerate_validation_phrase!(params, self)
        end
      end
    end
  end
end
